javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -783,7 +824,51 @@ public boolean isTraversableAgain() {
 
     @Override
     public Iterator<T> iterator() {
-        return isEmpty() ? Iterator.empty() : trie.iterator();
+        return isEmpty() ? Iterator.empty()
+                         : new VectorIterator();
+    }
+
```

Comment:
I've thought about it. The time-penalty of the simple, combined Iterator may come from the delegated Iterator calls:

* first the combined Iterator is called
* then the combined Iterator delegates to the current underlying Iterator

When I get some time (hopefully tomorrow) I will benchmark it, comparing a simple Iterator and a combined Iterator that delegates to a simple Iterator.